### PR TITLE
Feature: Data source (#58550)

### PR DESCRIPTION
**Pull request by nx-mukesh**
#Ticket
https://redmine.successive.tech/issues/58550

Description: 
Add Data Sources for fetching Data from Services, and using data source create mutation login User which will return the token of the user.

### CODE REVIEW

File: apollo-training/src/config/configurations.js [hunk 1]
```diff
@@ -10,6 +10,9 @@ const envVarsSchema = Joi.object({
     .default('development'),
   PORT: Joi.number()
     .default(9001),
+  SERVICE_URL: Joi.string()
```
Comment (permil-successive) [inline apollo-training/src/config/configurations.js:13]: What is JOI

File: apollo-training/src/config/configurations.js [hunk 2]
```diff
@@ -10,6 +10,9 @@ const envVarsSchema = Joi.object({
...
+    .default('http://localhost:9000'),
```
Comment (permil-successive) [inline apollo-training/src/config/configurations.js:14]: Don't Hardcode url

File: apollo-training/src/datasource/User.js [hunk 3]
```diff
@@ -1,19 +1,25 @@
...
   }
 
-  getMe(){
+  getMe() {
     return this.get('/me');
   }
 
-  loginUser(payload){
+  async loginUser(payload) {
     return this.post('/login', payload);
```
Comment (permil-successive) [inline apollo-training/src/datasource/User.js:15]: why adding async

File: apollo-training/src/modules/trainee/types.graphql [hunk 4]
```diff
@@ -1,5 +1,5 @@
-input CreateUserInput {
-  name: String!
-  email: String!
-  role: String!
+input CreateUserInput{
+    name: String!
+    email: String!
```
Comment (permil-successive) [inline apollo-training/src/modules/trainee/types.graphql:3]: why adding space

File: apollo-training/src/modules/user/mutation.js [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
-  }
-}
\ No newline at end of file
+  loginUser: async (parent, args, context) => {
+    const { payload: { email, password } } = args;
+    // console.log('args ---> ',args);
+    const { dataSources: { userAPI } } = context;
+    // console.log('UserAPI context ---> ',context);
+    const response = await userAPI.loginUser({ email, password });
+    console.log('response---> ', response);
```
Comment (permil-successive) [inline apollo-training/src/modules/user/mutation.js:19]: remove console

File: apollo-training/src/modules/user/query.js [hunk 6]
```diff
@@ -1,10 +1,21 @@
...
-      name: 'Mukesh Maurya',
-      email: 'mukesh.maurya@successive.tech',
-      role: 'trainee',
-    };
+  getMyProfile: async (parent, args, context) => {
+    // console.log('context ---> ',context);
+    // const { dataSources: { userAPI } }= context;
+    // console.log('context ---> ',context);
+    const response = await context.dataSources.userAPI.getMe();
+    console.log('response for get my profile query ---> ', response.data[0].email);
```
Comment (permil-successive) [inline apollo-training/src/modules/user/query.js:18]: remove console

File: apollo-training/src/modules/user/types.graphql [hunk 7]
```diff
@@ -1,8 +1,19 @@
-type User {
-  id: ID!
-  name: String!
-  email: String!
-  role: String!
+type User{
+    id:ID!
+    name: String!
```
Comment (permil-successive) [inline apollo-training/src/modules/user/types.graphql:3]: why adding space here

File: apollo-training/src/config/configurations.js (hunk 1 above)
Comment (nx-mukesh) [reply apollo-training/src/config/configurations.js:13]: Joi is a validator package, we create a schema for field and by joi we can validate request body. 
like Email: Joi.email().ToLowercase().required() and 
      Password: Joi.min(8).max(20).required()   etc

File: apollo-training/src/modules/user/types.graphql (hunk 7 above)
Comment (nx-mukesh) [reply apollo-training/src/modules/user/types.graphql:3]: fixed

File: apollo-training/src/modules/user/query.js (hunk 6 above)
Comment (nx-mukesh) [reply apollo-training/src/modules/user/query.js:18]: fixed

File: apollo-training/src/modules/user/mutation.js (hunk 5 above)
Comment (nx-mukesh) [reply apollo-training/src/modules/user/mutation.js:19]: fixed

File: apollo-training/src/modules/trainee/types.graphql (hunk 4 above)
Comment (nx-mukesh) [reply apollo-training/src/modules/trainee/types.graphql:3]: fixed

File: apollo-training/src/datasource/User.js (hunk 3 above)
Comment (nx-mukesh) [reply apollo-training/src/datasource/User.js:15]: because the function return promises so we are using new syntax async-await.

**Comment by permil-successive** [comment]
Approved

**Pull request merged**